src/progress.rs

```diff
@@ -167,60 +290,109 @@ impl ProgressSet {
+        debug!("Inserting voter with id {}.", id);
+
+        if self.learner_ids().contains(&id) {
+            return Err(Error::Exists(id, "learners"));
+        } else if self.voter_ids().contains(&id) {
             return Err(Error::Exists(id, "voters"));
+        } else if self.is_in_membership_change() {
+            return Err(Error::ViolatesContract(
+                "There is a pending membership change.".into(),
+            ));
         }
+
         self.configuration.voters.insert(id);
         self.progress.insert(id, pr);
+
         self.assert_progress_and_configuration_consistent();
         Ok(())
     }
 
-    /// Adds a learner to the cluster
-    pub fn insert_learner(&mut self, id: u64, pr: Progress) -> Result<(), Error> {
-        // If the progress exists already this is in error.
-        if self.progress.contains_key(&id) {
-            // Determine the correct error to return.
-            if self.learner_ids().contains(&id) {
-                return Err(Error::Exists(id, "learners"));
-            }
+    /// Adds a learner to the group.
+    ///
+    /// # Errors
+    ///
+    /// * `id` is in the voter set.
+    /// * `id` is in the learner set.
+    /// * There is a pending membership change.
+    pub fn insert_learner(&mut self, id: u64, pr: Progress) -> Result<()> {
+        debug!("Inserting learner with id {}.", id);
+
+        if self.learner_ids().contains(&id) {
+            return Err(Error::Exists(id, "learners"));
+        } else if self.voter_ids().contains(&id) {
             return Err(Error::Exists(id, "voters"));
+        } else if self.is_in_membership_change() {
+            return Err(Error::ViolatesContract(
+                "There is a pending membership change.".into(),
+            ));
         }
+
         self.configuration.learners.insert(id);
         self.progress.insert(id, pr);
+
         self.assert_progress_and_configuration_consistent();
         Ok(())
     }
 
     /// Removes the peer from the set of voters or learners.
-    pub fn remove(&mut self, id: u64) -> Option<Progress> {
-        self.configuration.voters.remove(&id);
+    ///
+    /// # Errors
+    ///
+    /// * There is a pending membership change.
+    pub fn remove(&mut self, id: u64) -> Result<Option<Progress>> {
+        debug!("Removing peer with id {}.", id);
+
+        if self.is_in_membership_change() {
+            return Err(Error::ViolatesContract(
+                "There is a pending membership change.".into(),
+            ));
+        }
+
         self.configuration.learners.remove(&id);
+        self.configuration.voters.remove(&id);
         let removed = self.progress.remove(&id);
+
         self.assert_progress_and_configuration_consistent();
-        removed
+        Ok(removed)
     }
 
     /// Promote a learner to a peer.
-    pub fn promote_learner(&mut self, id: u64) -> Result<(), Error> {
+    pub fn promote_learner(&mut self, id: u64) -> Result<()> {
```

Comment:
Docs should probably contain errors cases like the others methods.

Reply:
Let's leave this for future work, it's not directly relevant to this PR and this PR has been held up long enough.

I agree though. :)